src/transformers/integrations/hub_kernels.py

```diff
@@ -56,6 +60,36 @@
 
     register_kernel_mapping(_KERNEL_MAPPING)
 
+    def use_kernel_forward_from_hub(*args, **kwargs):
+        """
+        Expands `kernels`' `use_kernel_forward_from_hub` to NOT use a kernel at compile time. This should be removed
+        when `kernels` supports `torch.compile`.
+        """
+
+        def decorator_with_compile_path(cls):
+            # Keeps a reference to the original forward method
+            original_forward = cls.forward
+
+            # Applies the original decorator
+            decorator = original_use_kernel_forward_from_hub(*args, **kwargs)
+            cls = decorator(cls)
+
+            # Replaces the kernel forward with a compile-friendly version
+            kernel_forward = cls.forward
+
+            def forward_with_compile_path(*forward_args, **forward_kwargs):
+                if is_torchdynamo_compiling():
+                    return original_forward(*forward_args, **forward_kwargs)
+                else:
+                    return kernel_forward(*forward_args, **forward_kwargs)
```

Comment:
Maybe not super clean, but we can have an extra kwarg, smth like that 

```suggestion
            def forward_with_compile_path(*forward_args, **forward_kwargs):
                use_kernel = forward_kwargs.pop("use_kernel", True)
                if is_torchdynamo_compiling() or not use_kernel:
                    return original_forward(*forward_args, **forward_kwargs)
                else:
                    return kernel_forward(*forward_args, **forward_kwargs)
```

Reply:
Then, I can manage it out of the module call

Reply:
Yeah, good idea, that could work!

Reply:
@qubvel it hits a similar barrier: many layers don't receive `**kwargs`, so they will not be reached by arbitrary kwargs as in e.g. `model(..., use_kernel=False)` :(

Reply:
Custom backends https://pytorch.org/docs/stable/torch.compiler_custom_backends.html could be way to send exceptions at compile time, but probably a huge struggle - just posting here in case but I don't really believe in it

Reply:
ALthough actually registering a custom backend simply calling the user's backend + raising in case of kernels activated could actually work? 🤔

Reply:
@gante but they should not receive `**kwargs` actually, because we pop it out anyway before passing kwargs to original/kernel forward function

Reply:
I can do the following in the molding code later (just as example

```python
def forward(...):
    x = self.module_with_kernel(x, use_kerenel=not self.config.disable_custom_kernels)
```

Reply:
maybe we should name the argument `disable_custom_kernels` for consistency 😅 

```
def forward(...):
    x = self.module_with_kernel(x, disable_custom_kernels=self.config.disable_custom_kernels)
```

Reply:
or at least `disable_kernel`

Reply:
@qubvel what I meant is that in some layers we can't pipe the custom kwarg argument all the way from `model.forward` to the layer's `forward`, because the outer doesn't pass it. Likewise, at initialization time, some layers don't take  `**kwargs` nor `config`.

`Phi3RMSNorm` is one of such examples: it has a kernel, the outer layer doesn't pass `**kwargs` to `forward`, and at initialization time there are no `**kwargs` nor `config`

Reply:
Thanks for clarifying. I understand your point, but do we need to propagate this from the top module itself? We can still control it with the config, but some changes to the modeling code would be required.

For Phi3, `Phi3RMSNorm` does not accept the config, but we will control it from `Phi3DecoderLayer.forward`, which does accept and store `config`.

https://github.com/huggingface/transformers/blob/f9742143538cfeb65c3a25632b83276266819c41/src/transformers/models/phi3/modeling_phi3.py#L326

should be like:
```python
hidden_states = self.mlp(hidden_states, disable_kernel=self.config.disable_custom_kernels)  
```

But that's actually not required to be like that, and it's more of an exception. I suppose most cases would be covered with `is_torchdynamo_compiling`, and maybe deformable attention is the only kernel that will require such manual control. Alternatively, we can require for "kernel" modules to store config and check "disable_custom_kernels" flag in the decorator.

As I mentioned earlier, not super clean approach IMO, but should do the job

Reply:
@Cyrilvallez from a quick glance at the docs I still don't know how to raise exceptions if the user is compiling 👀 Quite possibly because I'm noob at this :D

I tried this
```py
    def my_custom_backend(gm, example_inputs):
        return gm.forward

    def use_kernel_forward_from_hub(*args, **kwargs):

        def decorator_with_compile_path(cls):
            # Applies the original decorator
            decorator = original_use_kernel_forward_from_hub(*args, **kwargs)
            cls = decorator(cls)

            # Replaces the kernel forward with a compile-friendly version
            kernel_forward = cls.forward

            def forward_with_compile_path(*forward_args, **forward_kwargs):
                if is_torchdynamo_compiling():
                    raise RuntimeError("Foo")
                else:
                    return kernel_forward(*forward_args, **forward_kwargs)

            cls.forward = torch.compile(forward_with_compile_path, backend=my_custom_backend)
            return cls

        return decorator_with_compile_path
```
(raising an exception in `my_custom_backend` doesn't work either)

Reply:
@qubvel I see your point: we leave the option to pass the flag, some layers won't do it by default, but at least we have some degree of control 👍

Reply:
@qubvel added ✅ I went with the config route, since most layers have it and it doesn't require forwarding any argument, but let me know if you'd prefer having control through an argument.

(sorry for being obtuse in the comments above: I was too focused on having a solution that worked in all cases, and completely missed the point that partial support would still be useful 😅 )

Reply:
Okay, thanks! I just realized neither of the options solves my issue entirely, but at least it will respect the configuration parameter -> the current implementation is going to work. However, for the refactored version with CoreML export support, I want to know in advance which path is going to be executed before passing inputs into the module. I have no idea how to get this, previously, it was resolved the following way:

```py
# kernel path
if kernel_loaded and not is_compiling and not custom_kernels_disabled:
    hidden_state = kernel_forward(hidden_state)  # guaranteed to be kernel forward

# eager path, avoid 6D tensor
else:
    hidden_state = hidden_state.reshape(...)
    hidden_state = kernel(hidden_state)

```

I will solve it when return to RT_DETR refactoring, and maybe by that time `kernels` library will be updated, not an urgent issue

Reply:
Good! We can revisit this later anyways, if we find a better way or new issues 🤗

Reply:
Ha yes, my idea was to raise directly in the custom backend if kernels are installed and the env variable is active, and force the use of that backend everywhere (thus not changing the kernel decorator). Too bad if it does not work!